### ISSUE TITLE
InvalidOperationException: An attempt was made to transition a task to a final state when it had already completed

### ISSUE THREAD

**Issue by PNicolle** [MicrosoftEdge/WebView2Feedback#382]
I'm using the '0.9.579-prerelease' version of the 'Microsoft.Web.WebView2' NuGet package in a x64bit WPF application.  This application is using AvalonDock as docking system.  The WebView2 host window is initialy tabbed then it is made floating. When trying to tab it again the following exception occurs:

System.InvalidOperationException : 'An attempt was made to transition a task to a final state when it had already completed.'
 at System.Threading.Tasks.TaskCompletionSource`1.SetResult(TResult result)
   at Microsoft.Web.WebView2.Wpf.WebView2.BuildWindowCore(HandleRef hwndParent)
   at System.Windows.Interop.HwndHost.BuildWindow(HandleRef hwndParent)
   at System.Windows.Interop.HwndHost.BuildOrReparentWindow()
   at System.Windows.Interop.HwndHost.OnSourceChanged(Object sender, SourceChangedEventArgs e)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.PresentationSource.UpdateSourceOfElement(DependencyObject doTarget, DependencyObject doAncestor, DependencyObject doOldParent)
   at System.Windows.PresentationSource.RootChanged(Visual oldRoot, Visual newRoot)
   at System.Windows.Interop.HwndSource.set_RootVisualInternal(Visual value)
   at System.Windows.Window.SetRootVisual()
   at System.Windows.Window.SetRootVisualAndUpdateSTC()
   at System.Windows.Window.SetupInitialState(Double requestedTop, Double requestedLeft, Double requestedWidth, Double requestedHeight)
   at System.Windows.Window.CreateSourceWindow(Boolean duringShow)
   at System.Windows.Window.ShowHelper(Object booleanBox)
   at System.Windows.Window.ShowDialog()
   at SoftwareFoundationSDK.Application.<>c__DisplayClass2_0.<OnDefaultAppDomainUnhandledException>b__0() in C:\Git\V8_5Avalon\HybridApp\HybridApp.Workbench\SfSdk.Workbench\Application.xaml.cs:line 92

Any idea of the root cause and how to solve it?
Thanks in advance.

[AB#28321216](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/28321216)

**Comment by pagoe-msft** [comment]
@PNicolle ,

I've filed a bug for our team to investigate.

**Comment by AndyT-MS** [comment]
Hi @PNicolle,

I'm working on fixing this bug, but I'm having trouble reproducing it.  I'm pretty sure that this bug will generally occur anytime the `WebView2` is moved to a different `PresentationSource` (i.e. parent `HWND`), but my attempts to cause that to happen have been unsuccessful.

Could you please confirm which version of AvalonDock you're using?  I'm not familiar with that library, but it looks like there's a proprietary one (Xceed) and an [open source one](https://github.com/Dirkster99/AvalonDock) which appear to have a shared ancestry.  I downloaded the open source one and hacked a `WebView2` control into a new tool window in its "TestApp" project, but when I pull that tool window out into an independent floating window and then drag it back into a tab it works fine and doesn't throw any exception.

Any chance you could share a minimal repro app?

Thanks!

**Comment by PNicolle** [comment]
Hi Andy,

We are using the version 3.5 of the open source library. But as we detected several bugs in it, we made some fixes. Also we use a custom theme. This could explain why  you cannot reproduce with the TestApp...
I could share privately a repro app, but the architecture is quite complex because is based on an internal framework we developped. Anyway if you want it, tell me how I can send it to you.

Regards.

**Comment by PNicolle** [comment]
FYI, I switched to the version "0.9.628-prerelease" and have Microsoft Edge Canary Version 88.0.672.0  (64 bits).  With this environment, the InvalidOperationException does not occur anymore :-)
Did you fix it specifically or it is fixed by accident?

Regards.

**Comment by champnic** [comment]
Sounds like it was fixed by accident by changes between 579 and 628. Not ideal (we prefer to understand what's going wrong) but good news nonetheless! I'm going to close this, but please let me know if you run into it again and we can reopen. Thanks!

**Comment by AndyT-MS** [comment]
Sorry to be so slow to write back here.  Although I was never able to repro this, I did actually make some specific changes in an attempt to address it.  I had been planning to notify you to try them out once the new prerelease package was available, but I ended up having to focus on completely unrelated matters for effectively all of October.  Thanks for checking up on it anyway and letting us know how it went.  I'm really glad to hear that my changes seem to have fixed it!
